contracts/eosio.system/src/powerup.cpp

```diff
@@ -0,0 +1,398 @@
+      state.decay_secs           = *args.decay_secs;
+      state.min_price            = *args.min_price;
+      state.max_price            = *args.max_price;
+   };
+
+   if (!args.powerup_days) {
+      *args.powerup_days = state.powerup_days;
+   }
+
+   if (!args.min_powerup_fee) {
+      eosio::check(!is_default_asset(state.min_powerup_fee), "min_powerup_fee does not have a default value");
+      *args.min_powerup_fee = state.min_powerup_fee;
+   }
+
+   eosio::check(*args.powerup_days > 0, "powerup_days must be > 0");
+   eosio::check(args.min_powerup_fee->symbol == core_symbol, "min_powerup_fee doesn't match core symbol");
+   eosio::check(args.min_powerup_fee->amount > 0, "min_powerup_fee must be positive");
+
+   state.powerup_days    = *args.powerup_days;
+   state.min_powerup_fee = *args.min_powerup_fee;
+
+   update(state.net, args.net);
+   update(state.cpu, args.cpu);
+
+   update_weight(now, state.net, net_delta_available);
+   update_weight(now, state.cpu, cpu_delta_available);
+   eosio::check(state.net.weight >= state.net.utilization, "weight can't shrink below utilization");
+   eosio::check(state.cpu.weight >= state.cpu.utilization, "weight can't shrink below utilization");
+   state.net.adjusted_utilization = std::min(state.net.adjusted_utilization, state.net.weight);
+   state.cpu.adjusted_utilization = std::min(state.cpu.adjusted_utilization, state.cpu.weight);
+
+   adjust_resources(get_self(), reserv_account, core_symbol, net_delta_available, cpu_delta_available, true);
+   state_sing.set(state, get_self());
+} // system_contract::configpower
+
+/**
+ *  @pre 0 <= state.min_price.amount <= state.max_price.amount
+ *  @pre 0 < state.max_price.amount
+ *  @pre 1.0 <= state.exponent
+ *  @pre 0 <= state.utilization <= state.adjusted_utilization <= state.weight
+ *  @pre 0 <= utilization_increase <= (state.weight - state.utilization)
+ */
+int64_t calc_powerup_fee(const powerup_state_resource& state, int64_t utilization_increase) {
+   if( utilization_increase <= 0 ) return 0;
+
+   // Let p(u) = price as a function of the utilization fraction u which is defined for u in [0.0, 1.0].
+   // Let f(u) = integral of the price function p(x) from x = 0.0 to x = u, again defined for u in [0.0, 1.0].
+
+   // In particular we choose f(u) = min_price * u + ((max_price - min_price) / exponent) * (u ^ exponent).
+   // And so p(u) = min_price + (max_price - min_price) * (u ^ (exponent - 1.0)).
+
+   // Returns f(double(end_utilization)/state.weight) - f(double(start_utilization)/state.weight) which is equivalent to
+   // the integral of p(x) from x = double(start_utilization)/state.weight to x = double(end_utilization)/state.weight.
+   // @pre 0 <= start_utilization <= end_utilization <= state.weight
+   auto price_integral_delta = [&state](int64_t start_utilization, int64_t end_utilization) -> double {
+      double coefficient = (state.max_price.amount - state.min_price.amount) / state.exponent;
```

Comment:
`state.exponent` will be set to `2.0` if cfgpowerup does not receive an exponent in the config parameters.

Reply:
`weight` cannot be zero because of this line: https://github.com/EOSIO/eosio.contracts/blob/faf3d43bb70dab6d52b5b4979a38b0625b85344a/contracts/eosio.system/src/powerup.cpp#L359